### PR TITLE
Fix Relative class names are not converted to fully-qualified class names (FQCNs)

### PR DESCRIPTION
**Pull request by SavKS**
Fixes https://github.com/barryvdh/laravel-ide-helper/issues/627

### CODE REVIEW

**Comment by mfn** [comment]
For the lazy: https://github.com/barryvdh/laravel-ide-helper/issues/627

Would be great to receive some tests for this!

**Review by mfn — CHANGES_REQUESTED** [review]
@SavKS there's a lot of ground to cover the feature sounds super-interesting!

We haven't heard back, are you still interested in finish the PR?

I think due to the complexity, some tests are definitely on order here, what do you think?

**Comment by SavKS** [comment]
@mfn 

I'm sorry, but I can't write tests. If anyone in the community has the opportunity to do so, I would be grateful for that.

**Comment by mfn** [comment]
Thanks for the update!

This changes are quite involved I'm leaning towards having tests is a must.

I don't intend to do this myself, so let's keep the PR open if someone volunteers to step up!

**Comment by SeinopSys** [comment]
I spend a good two hours trying to write tests for this but it seems like the `GeneratorCommand` class uses some special magic to detect facades which I was not able to replicate. Below is my unsuccessful attempt in case someone is able to build on it in any capacity.

```diff
Index: src/Method.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/Method.php	(revision 39e2e8e2bcade1c9fbd2f5a3da7ea7f105c20206)
+++ src/Method.php	(date 1601068825567)
@@ -46,7 +46,7 @@
      * @param \ReflectionClass $class
      * @param string|null $methodName
      * @param array $interfaces
-     * @param NamespaceUses $namespaceUses
+     * @param NamespaceUses|null $namespaceUses
      */
     public function __construct(
         $method,
Index: src/Console/GeneratorCommand.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/Console/GeneratorCommand.php	(revision 39e2e8e2bcade1c9fbd2f5a3da7ea7f105c20206)
+++ src/Console/GeneratorCommand.php	(date 1601060442663)
@@ -58,11 +58,9 @@
      * @param \Illuminate\View\Factory $view
      */
     public function __construct(
-        /*ConfigRepository */
-        $config,
+        \Illuminate\Config\Repository $config,
         Filesystem $files,
-        /* Illuminate\View\Factory */
-        $view
+        \Illuminate\View\Factory $view
     ) {
         $this->config = $config;
         $this->files = $files;
Index: tests/TestCase.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/TestCase.php	(revision 39e2e8e2bcade1c9fbd2f5a3da7ea7f105c20206)
+++ tests/TestCase.php	(date 1601064080820)
@@ -65,6 +65,15 @@
                 return file_get_contents($file);
             });
 
+        $mockFilesystem
+            ->shouldReceive('exists')
+            ->andReturnUsing(function ($file) {
+                return file_exists($file);
+            });
+
+        $mockFilesystem
+            ->shouldReceive('getRequire');
+
         $mockFilesystem
             ->shouldReceive('put')
             ->with(
Index: tests/Console/GeneratorCommand/RelativeClassNames/Test.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/Console/GeneratorCommand/RelativeClassNames/Test.php	(date 1601068862904)
+++ tests/Console/GeneratorCommand/RelativeClassNames/Test.php	(date 1601068862904)
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Barryvdh\LaravelIdeHelper\Tests\Console\GeneratorCommand\RelativeClassNames;
+
+use Barryvdh\LaravelIdeHelper\IdeHelperServiceProvider;
+use Barryvdh\LaravelIdeHelper\Tests\Console\GeneratorCommand\RelativeClassNames\ExternalDependency\ExternalServiceProvider;
+use Barryvdh\LaravelIdeHelper\Tests\SnapshotPhpDriver;
+use Barryvdh\LaravelIdeHelper\Tests\TestCase;
+use Illuminate\Support\Facades\View;
+use Illuminate\View\FileViewFinder;
+
+class Test extends TestCase {
+    /** @var string */
+    private $cwd;
+
+    /** @var FileViewFinder */
+    private $viewFinder;
+
+    protected function setUp():void {
+        parent::setUp();
+
+        new ExternalServiceProvider($this->app);
+
+        $this->cwd = getcwd();
+
+        $this->viewFinder = View::getFinder();
+
+        $this->mockFilesystem();
+
+        View::setFinder(new FileViewFinder(app()->get('files'), [
+            realpath(__DIR__).'/../../../../resources/views',
+        ]));
+    }
+
+    protected function tearDown():void {
+        chdir($this->cwd);
+
+        View::setFinder($this->viewFinder);
+
+        parent::tearDown();
+    }
+
+    public function testCommand():void {
+        chdir(__DIR__);
+
+        $result = $this->artisan('ide-helper:generate');
+
+        $result->assertExitCode(0);
+        $this->assertMatchesSnapshot($this->mockFilesystemOutput, new SnapshotPhpDriver());
+    }
+
+    /**
+     * Get package providers.
+     *
+     * @param \Illuminate\Foundation\Application $app
+     *
+     * @return array
+     */
+    protected function getPackageProviders($app) {
+        return [IdeHelperServiceProvider::class];
+    }
+}
Index: src/NamespaceUses.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/NamespaceUses.php	(revision 39e2e8e2bcade1c9fbd2f5a3da7ea7f105c20206)
+++ src/NamespaceUses.php	(date 1601068825567)
@@ -30,7 +30,7 @@
     /**
      * @var array
      */
-    public $classAliases;
+    public $classAliases = [];
 
     /**
      * NamespaceUses constructor.
@@ -60,11 +60,17 @@
     {
         $parser = (new ParserFactory())->create(ParserFactory::PREFER_PHP7);
 
+        $fileName = (new \ReflectionClass($this->classFQN))->getFileName();
+
+        if (!file_exists($fileName)) {
+            return;
+        }
+
         /** @var Namespace_ $namespaceData */
         $namespaceData = Arr::first(
             $parser->parse(
                 file_get_contents(
-                    (new \ReflectionClass($this->classFQN))->getFileName()
+                    $fileName
                 )
             ),
             function ($node) {
```

**Comment by mfn** [comment]
:-(

I made a note to take a look sometime later, thanks!

**Comment by mfn** [comment]
> I made a note to take a look sometime later, thanks!

I did make a note but coming back to this, I don't feel like having the energy going through this without any test coverage.

I'm happy to assist anyone else taking a stab a this tough, just create a new PR with this commit and hack away and ping me. It's easier to talk about this in PRs with actual code then with sharing ZIPs.

**Comment by mfn** [comment]
It looks like with the recent update to this PR something went wrong with a merge? I should not see so many unrelated changes here.

Please fix this and update with master and resolve the merge conflicts, thank you

**Review by barryvdh — DISMISSED** [review]

**Review by mfn — COMMENTED** [review]
Personally I still don't feel confident merging this with 0 tests

**Comment by SavKS** [comment]
Tests were added with the help of @Tarik02.

File: src/UsesResolver.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * @param string $classFQN
+     * @param string|null $alias
+     * @return void
+     */
+    protected function addClassAlias(string $classFQN, string $alias = null)
+    {
+        if ($alias === null) {
+            $alias = class_basename($classFQN);
```
Comment (mfn) [inline src/UsesResolver.php:129]: Unfortunately can't use global functions from Support/helpers.php in a library.

ide-helper, whilst Laravel specific, may be used just with its individual illuminate/* packages and must not depend on such functions.

Fix in this case: sorry, just copypaste the code over 😬

File: src/UsesResolver.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+use PhpParser\Node\Stmt\Use_;
+use PhpParser\Node\Stmt\UseUse;
+use PhpParser\ParserFactory;
+
+class UsesResolver
+{
+    /**
+     * @var array
+     */
+    protected $classAliases = [];
```
Comment (mfn) [inline src/UsesResolver.php:26]: From a design perspective of this class, I suggest to make it stateless as there's no benefit to it.

Suggest approach:
- get rid of this property
- just return the data array from `loadFromCode`
- get rid of `addClassAlias
- change all the `load*` methods to just return the data you return from `loadFromCode`
- which means the call in `\Barryvdh\LaravelIdeHelper\Alias::__construct` is just `(new UsesResolver())->loadFromClass($this->root);`
- adapt test

Most classes in this library have their fair share of states, which is nightmare to trace properly.

**Review by mfn — CHANGES_REQUESTED** [review]
Approach LGTM and tested it also on a real life project 🎉 

Some technical nitpicks, please see feedback inline

**Review by mfn — APPROVED** [review]
LGTM!

If you could add a README entry, I think we've good to go!

**Comment by mfn** [comment]
> If you could add a README entry, I think we've good to go!

PS: I meant changelog of course -> already made a PR https://github.com/barryvdh/laravel-ide-helper/pull/1176

**Pull request merged**